working/sealed-types/feature-specification.md

```diff
@@ -0,0 +1,196 @@
+# Sealed types
+
+Author: Bob Nystrom
+
+Status: In-progress
+
+Version 1.0
+
+This proposal specifies *sealed types*, which is core capability needed for
+[exhaustiveness checking][] of subtypes in [pattern matching][]. This proposal
+is a subset of the [type modifiers][] proposal. (We may wish to do all or parts
+of the rest of that proposal, but they aren't needed for pattern matching, so
+this proposal separates them out.) For motivation, see the previously linked
+documents.
+
+[exhaustiveness checking]: https://github.com/dart-lang/language/blob/master/working/0546-patterns/exhaustiveness.md
+
+[pattern matching]: https://github.com/dart-lang/language/blob/master/working/0546-patterns/patterns-feature-specification.md
+
+[type modifiers]: https://github.com/dart-lang/language/blob/master/working/type-modifiers/feature-specification.md
+
+## Introduction
+
+Marking a type `sealed` applies two restrictions:
+
+*   If it's a class, the type itself can't be directly constructed. The class is
+    implicitly `abstract`.
+
+*   All direct subtypes of the type must be defined in the same library. Any
+    types that directly implement, extend, or mixin the sealed type must be
+    defined in the library where the sealed type is defined.
+
+In return for those restrictions, sealed provides two useful properties for
+exhaustiveness checking:
+
+*   All of the direct subtypes of the sealed type can be easily found and
+    enumerated.
+
+*   Any concrete instance of the sealed type must also be an instance of at
+    least one of the known direct subtypes. In other words, if you match on a
+    value of the sealed type and you have cases for all of the direct subtypes,
+    the compiler knows those cases are exhaustive.
+
+### Open subtypes
+
+Note that it is *not* necessary for the subtypes of a sealed type to themselves
+be sealed or closed to subclassing or implementing. Given:
+
+```dart
+sealed class Either {}
```

Comment:
Not sold on `sealed` not meaning transitively sealed.
I think people will expect `sealed` to mean that nobody can create instances implementing the type, not just that nobody can create *immediate* subtypes.

Reply:
> Not sold on `sealed` not meaning transitively sealed.

This is consistent with how `sealed` works [in Kotlin](https://kotlinlang.org/docs/sealed-classes.html#location-of-direct-subclasses) [and Java](https://www.baeldung.com/java-sealed-classes-interfaces#2-sealed-classes) (which does require you to explicit opt the subclass in by marking it `non-sealed`, but the capability is there).

>  not just that nobody can create immediate subtypes.

If you want that, you can always mark the subtypes `sealed` too. (Or `closed`/`base` if we get those.)

Fundamentally, the way I've approached this design is, "What is the simplest set of restrictions that we need for exhaustiveness checking to work." As far as pattern matching cares, if you have a sealed type `Either` and two subtypes `Left` and `Right`, it does not matter at all if `Left` or `Right` happen to have their own subtypes. It's perfectly sound either way. Given that, I don't see any reason to add an unneeded restriction (which users can opt into already if they want).